Implement wxAuiNotebook::HitTest

walderich opened the pull request:
As discussed in Trac ticket #18290, I tried to implement HitTest for wxAuiNotebook. Unfortunately I discovered that changing `GetTabCtrlFromPoint` to `const` doesn't help a lot, so I didn't include any changes there.

I am not sure how to check for a label or image hit. Therefore I only implemented the NOWHERE, ONPAGE and ONITEM flags.

I also used the "worst case" `const_cast`, as I didn't want to introduce a new interface on wxAuiManager just for the `HitTest` interface. Any suggestions replacing the cast are welcome and I will change it accordingly.

walderich committed:
Implement wxAuiNotebook::HitTest result

Check if point is on window or a certain tab.

walderich committed:
Serve wxAuiNotebook::HitTest flags

Add support for flags wxBK_HITTEST_NOWHERE, wxBK_HITTEST_ONPAGE and
wxBK_HITTEST_ONITEM.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3359,10 +3359,35 @@ void wxAuiNotebook::SetPageSize (const wxSize& WXUNUSED(size))
-    return wxNOT_FOUND;
+    wxWindow *w = NULL;
+    long position = wxBK_HITTEST_NOWHERE;
+    wxAuiPaneInfoArray& all_panes = const_cast<wxAuiManager&>(m_mgr).GetAllPanes();
```
Can this reference be made const or do we need to call non-const methods on it too?

walderich replied:
Yes, that's indeed possible. I updated this on the branch. As this is my first PR: Can I update this PR with the new changes, or do I need to create a new one?

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3359,10 +3359,35 @@ void wxAuiNotebook::SetPageSize (const wxSize& WXUNUSED(size))
+    wxWindow *w = NULL;
+    long position = wxBK_HITTEST_NOWHERE;
+    wxAuiPaneInfoArray& all_panes = const_cast<wxAuiManager&>(m_mgr).GetAllPanes();
+    size_t i, pane_count = all_panes.GetCount();
```
```suggestion
    const size_t pane_count = all_panes.GetCount();
```

walderich replied:
Changed this. I would have even moved `GetCount` in the loop condition, as I thought I once read that the compiler even optimizes this away, when it sees that the result won't change during the loop. But I actually copied the code from another function in the same file and thought that should be the way to go :-)

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3359,10 +3359,35 @@ void wxAuiNotebook::SetPageSize (const wxSize& WXUNUSED(size))
+    long position = wxBK_HITTEST_NOWHERE;
+    wxAuiPaneInfoArray& all_panes = const_cast<wxAuiManager&>(m_mgr).GetAllPanes();
+    size_t i, pane_count = all_panes.GetCount();
+    for (i = 0; i < pane_count; ++i)
```
```suggestion
    for (size_t i = 0; i < pane_count; ++i)
```
(to go with the suggestion above)

walderich replied:
Also changed this locally. I now saw that I can commit your changes, too. I will do that the next time ...

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3359,10 +3359,35 @@ void wxAuiNotebook::SetPageSize (const wxSize& WXUNUSED(size))
+                position = wxBK_HITTEST_ONITEM;
+            break;
+        }
+        else if (tabframe->m_rect.Contains(pt) && tabframe->m_rect.y + tabframe->m_tabCtrlHeight < pt.y)
```
I'm not sure why is the 2nd part of the test here is needed, wouldn't it be caught by the test above if it were not true?

Maybe a comment could clarify things here...

walderich replied:
You're absolutely right. I removed the second condition.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3359,10 +3359,35 @@ void wxAuiNotebook::SetPageSize (const wxSize& WXUNUSED(size))
+        else if (tabframe->m_rect.Contains(pt) && tabframe->m_rect.y + tabframe->m_tabCtrlHeight < pt.y)
+        {
+            w = tabframe->m_tabs->GetWindowFromIdx(tabframe->m_tabs->GetActivePage());
+            if (w) position = wxBK_HITTEST_ONPAGE;
```
Minor, but could you please put the `if` statement body on its own line? If nothing else, it makes it simple to put a breakpoint on it, unlike the current version. And IMO it's also much more readable.

walderich replied:
Done.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3359,10 +3359,35 @@ void wxAuiNotebook::SetPageSize (const wxSize& WXUNUSED(size))
+        }
+    }
+
+    if (flags) *flags = position;
```
Same as above.

walderich replied:
Done.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3359,10 +3359,35 @@ void wxAuiNotebook::SetPageSize (const wxSize& WXUNUSED(size))
+    }
+
+    if (flags) *flags = position;
+    return w? GetPageIndex(w): wxNOT_FOUND;
```
```suggestion
    return w ? GetPageIndex(w) : wxNOT_FOUND;
```
Even more minor, but it's more readable to use spaces and the existing code always uses them.

walderich replied:
Also changed this on the updated branch.

vadz reviewed (commented):
No real comments, ideal would be to have a unit test for this but it looks like we don't have any AUI unit tests at all and I don't know how easy, or difficult, would it be to add them.

Have you tested this in the notebook sample BTW (it has `Ctrl+H` shortcut allowing to call `HitTest()` for the mouse cursor position)?

walderich committed:
Beautify wxAuiNotebook::HitTest

Added newlines and spaces for readability/maintainability

walderich committed:
Define locals as "const"

Add "const" modifier to all internal variables that are not changed
during the function. And as result also reduce lifespan of loop
variable.

walderich committed:
Remove superfluous range check

Remove check for point being below tab area. This case is already
processed when checking the tab area.

walderich commented:
> ideal would be to have a unit test for this but it looks like we don't have any AUI unit tests at all and I don't know how easy, or difficult, would it be to add them.

I guess I can try to create one. I saw that there's a "tests" subdirectory with the different tests suites. But I never did this before. I found that CppUnit is used for this. Is there any documentation or tutorial I can use to get started?

> Have you tested this in the notebook sample BTW (it has Ctrl+H shortcut allowing to call HitTest() for the mouse cursor position)?

In fact I created a separate GUI for this with a periodic timer displaying the result of the current mouse position in the status bar. This worked correctly for all cases, also with one and more split tab frames. I re-tested this also after the changes introduced from your suggestions and I could verify that the result is the same (especially for the removed 2nd `if` condition).

vadz commented:
Thanks for the updates, looks good now and I think I can even merge this before 3.1.2 (due tomorrow).

As for the tests, the CppUnit macros are actually a lie, they're kept for compatibility but actually we use CATCH underneath, please see the [tests HOWTO](https://github.com/wxWidgets/wxWidgets/blob/master/docs/contributing/how-to-write-unit-tests.md) for more information. And any PRs adding tests are always welcome!
